### DISCUSSION TITLE
Extract Role Memberships & Roles from all models in a workspace

### DISCUSSION THREAD

**Discussion by richardappleton** [DaxStudio/DaxStudio#1239] [Q&A]
Hi there, please could someone help quite urgently? 

I need to extract a list of all ROLES and ROLE_MEMBERSHIPS for every model in a workspace, I know I can do this manually but I need to do this at scale for ~100 workspaces.  I'm familiar with PowerShell if anyone knows if this is a possibility, or even within DAX studio where I could export all of the results. 

I basically need the output from these two commands, but for all models in a workspace:
select * from $SYSTEM.TMSCHEMA_ROLE_MEMBERSHIPS
select * from $SYSTEM.TMSCHEMA_ROLES

Any help appreciated, 
Richard

**Comment by dgosbell** [comment]
DAX Studio runs queries against a single data model at a time. There are no looping mechanisms or ways of altering the connection from within a DAX query. This is good from a security perspective as it blocks a lot of the attacks that are possible via SQL injection, but it also blocks scenarios like yours. 

But you should be able to do this with Powershell, you would just need to use something like ADOMD.NET to query via the XMLA endpoint. You could use the Power BI REST API to get a list of all the workspaces, then use the either also use the REST API to get a list of datasets or use the select * from $SYSTEM.DBSCHEMA_CATALOGS DMV. Then you can pull the roles and memberships using the above DMV calls

**Comment by richardappleton** [comment]
Thanks very much for this information, I will look into it.
Kind regards, Richard

On Mon, 26 Feb 2024 at 00:22, Darren Gosbell ***@***.***>
wrote:

> DAX Studio runs queries against a single data model at a time. There are
> no looping mechanisms or ways of altering the connection from within a DAX
> query. This is good from a security perspective as it blocks a lot of the
> attacks that are possible via SQL injection, but it also blocks scenarios
> like yours.
>
> But you should be able to do this with Powershell, you would just need to
> use something like ADOMD.NET to query via the XMLA endpoint. You could
> use the Power BI REST API to get a list of all the workspaces, then use the
> either also use the REST API to get a list of datasets or use the select *
> from $SYSTEM.DBSCHEMA_CATALOGS DMV. Then you can pull the roles and
> memberships using the above DMV calls
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DaxStudio/DaxStudio/discussions/1239#discussioncomment-8586217>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYRQXLCODZQ4HYRZUGNFYSLYVPIWDAVCNFSM6AAAAABDSVRLRSVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DKOBWGIYTO>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 

Richard Appleton

Data & Analytics Engagement Manager

UK


​Kingswood House, Kingswood Crescent, Cannock, Staffordshire, WS11 8JB

www.veolia.co <http://www.veolia.com>.uk

<http://www.veolia.co.uk>

-- 





Legal Disclaimer on behalf of the following companies registered in 
England and Wales under the following numbers (and all subsidiaries and 
group companies of the same):

Veolia UK Limited (02664833); Veolia 
Environmental Services Group (UK) Limited (02215767); Veolia Water UK 
Limited (02127283); Veolia Energy UK Limited (00883131);

Registered office 
210 Pentonville Road, London N1 9JY.

 

The information in this email and 
any associated files is confidential and may be legally privileged. It may 
also contain information that is subject to copyright or constitutes a 
trade secret. It is intended solely for the named recipient. Access to this 
email by anyone else is unauthorised.

If you are not the intended 
recipient, please note that any use,disclosure, copying, distribution of 
this email or any action taken or omitted to be taken in reliance on it is 
prohibited.

Warning: Although this email and any attachments are believed 
to be free from viruses, it is the responsibility of the recipient to 
ensure that they are virus free. No responsibility is accepted by any of 
the Veolia group companies for any loss or damage arising in any way from 
their receipt or opening.
